---
title: "camera gene set analysis"
author: ""
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: yeti
    highlight: tango
    code_folding: show
    keep_md: true
editor_options: 
  chunk_output_type: console
references:
- id: Robinson2010edgeR
  title: edgeR-a Bioconductor package for differential expression analysis of digital gene expression data
  author:
  - family: Robinson
    given: Mark D
  - family: McCarthy
    given: Davis J
  - family: Smyth
    given: Gordon K
  container-title: Bioinformatics
  volume: 26
  page: 139-140
  type: article-journal
  URL: https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btp616
  issued:
    year: 2010
- id: Robinson2010TMM
  title: A scaling normalization method for differential expression analysis of RNA-seq data
  author:
  - family: Robinson
    given: Mark D
  - family: Oshlack
    given: Alicia
  container-title: Genome Biology
  volume: 11
  page: R25
  type: article-journal
  URL: https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25
  issued:
    year: 2010
- id: Soneson2016tximport
  title: Differential analyses for RNA-seq- transcript-level estimates improve gene-level inferences
  author:
  - family: Soneson
    given: Charlotte
  - family: Love
    given: Michael I
  - family: Robinson
    given: Mark D
  container-title: F1000Research
  volume: 4
  page: 1521
  type: article-journal
  URL: https://f1000research.com/articles/4-1521/v2
  issued:
    year: 2016
- id: Wu2012camera
  title: Camera- a competitive gene set test accounting for inter-gene correlation
  author:
  - family: Wu
    given: Di
  - family: Smyth
    given: Gordon K
  container-title: Nucleic Acids Research
  volume: 40
  page: e133
  type: article-journal
  issued:
    year: 2012
---

```{r camera-setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, dev = c("png", "pdf"))
```

# Introduction

Here, we perform gene set analysis with the camera method [@Wu2012camera] from 
the limma package, based on gene-level abundance estimates from Salmon. 
For more detailed 
information about the individual steps in the analysis, we refer to the 
[edgeR](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf) 
and [limma](https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf) 
user guides.

# Load packages

```{r camera-load-pkg}
suppressPackageStartupMessages({
  library(dplyr)
  library(tximport)
  library(tximeta)
  library(SummarizedExperiment)
  library(edgeR)
  library(ggplot2)
  library(msigdbr)
})
```

# Load `SummarizedExperiment` object

We load the `SummarizedExperiment` objects prepared using `tximeta`, containing
gene- and transcript-level counts and feature lengths. In this report, we will
use the gene-level quantifications.

```{r camera-print-se}
## List of SummarizedExperiment objects (gene/transcript level)
se

## Get gene-level SummarizedExperiment object
sg <- se$sg
metadata <- colData(sg)

sg
```

# Load gene sets ******** CONFIRM CATEGORIES ********

We will use `camera` to perform an enrichment analysis for a collection of 
gene sets from the [mSigDB](http://software.broadinstitute.org/gsea/msigdb), 
packaged in the `msigdbr` R package. Here, we load the gene set definitions 
and select which ones to include in the analysis.

```{r camera-load-genesets}
m_df = msigdbr(species = organism,
               category = c("H", "C5"))
```

# Create DGEList and include average transcript length offsets

The `DGEList` is the object type used by `edgeR` to store all the information 
related to an experiment. Here, we create a `DGEList` containing gene-level 
read counts as well as offsets derived from average transcript lengths 
[@Soneson2016tximport]. Normalization factors are calculated using the TMM 
method [@Robinson2010TMM]. We also include gene annotation information.

```{r camera-dge-generate}
## Extract count matrix and average transcript lengths
cts <- assay(sg, "counts")
normMat <- assay(sg, "length")
normMat <- normMat/exp(rowMeans(log(normMat)))

## Generate offset matrix
o <- log(edgeR::calcNormFactors(cts/normMat)) + 
  log(colSums(cts/normMat))

## Create DGEList object and add offsets
dge0 <- edgeR::DGEList(cts)
dge0 <- edgeR::scaleOffset(dge0, offset = t(t(log(normMat)) + o))
dge0 <- edgeR::calcNormFactors(dge0)

## Add gene annotation to the DGEList object
dge0$genes <- as.data.frame(rowRanges(sg))
```

# Define design. ******** DEFINE APPROPRIATE DESIGN ******** 

Next, we specify the design matrix of the experiment, defining which sample 
annotations will be taken into account in the statistical modeling.

```{r camera-define-design}
stopifnot(all(colnames(dge0) == metadata$names))
#(des <- model.matrix(~ XXXX, data = metadata))
#e.g.
(des <- model.matrix(~ 0 + celline, data = metadata))
```

# Filter out lowly expressed genes

We exclude lowly expressed genes from the subsequent analysis, since these 
contain little useful information for reliable statistical analysis. 

```{r camera-filter-genes}
dim(dge0)
keep <- edgeR::filterByExpr(dge0, design = des)
dge <- dge0[keep, ]
dge <- calcNormFactors(dge)
dim(dge)
```

# Estimate dispersion

Before performing the statistical analysis, we estimate the dispersion for each 
gene and plot the biological coefficient of variation against the estimated 
expression level. 

```{r camera-estimate-disp}
## Estimate dispersion and fit model
dge <- estimateDisp(dge, design = des)

## Plot dispersions
plotBCV(dge)
```

# Define contrasts ******** DEFINE PROPER CONTRASTS ******** 

Here we define the _contrasts_ to test by `camera`, i.e., the statistical 
comparisons of interest. For more information on defining proper contrasts, 
we refer to the [edgeR](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf) 
and [limma](https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf) 
user guides.


```{r camera-define-contrasts}
#(contrasts <- as.data.frame(makeContrasts(XXXX, levels = des)))
#e.g.
(contrasts <- as.data.frame(makeContrasts(
  contrasts = "cellineN61311-cellineN052611", levels = des
)))
```

# Perform tests ******** CONFIRM GENE SET SIZE LIMITS AND INTER.GENE.COR PARAMETER ********

Next, we perform the gene set analysis. We consider only gene sets where the 
number of genes shared with the data set is not too small and not too large. 
`camera` is a competitive gene set test that accounts for correlations among 
the genes within a gene set. 

```{r camera-perform-tests}
minSize <- 3
maxSize <- 500

## Get index for genes in each gene set in the DGEList
indexList <- limma::ids2indices(
  gene.sets = lapply(split(m_df, f = m_df$gs_name), function(w) w$gene_symbol),
  identifiers = dge$genes$gene_name,
  remove.empty = TRUE
)

## Filter out too small or too large gene sets
gsSizes <- vapply(indexList, length, 0)
indexList <- indexList[gsSizes >= minSize & gsSizes <= maxSize]

camera_res <- lapply(contrasts, function(cm) {
  camera(dge, index = indexList, design = des, contrast = cm, 
         inter.gene.cor = NA)
})
```

# Write results to text files

The results from `camera` are written to a separate text file for each tested 
contrast.

```{r camera-save-results}
## Write results to text files
if (is(camera_res, "data.frame")) {
  write.table(camera_res %>% tibble::rownames_to_column("GeneSet") %>%
                dplyr::arrange(PValue), 
              file = "camera_dge_results.txt", 
              sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)
} else {
  for (nm in names(camera_res)) {
    write.table(camera_res[[nm]] %>% 
                  tibble::rownames_to_column("GeneSet") %>%
                  dplyr::arrange(PValue), 
                file = paste0("camera_dge_results_", nm, ".txt"), 
                sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)
  }
}
```

The `camera` output, as well as the used gene sets, are saved to a file.

```{r camera-save-se}
geneSets <- lapply(indexList, function(i) dge$genes$gene_name[i])
saveRDS(list(cameraRes = camera_res,
             geneSets = geneSets), file = "camera_gsa.rds")
```

# Session info

The analyses above were performed with the following package versions:

```{r camera-session-info}
sessionInfo()
date()
```

# References