#!/usr/bin/env Rscript

args <- (commandArgs(trailingOnly = TRUE))
for (i in seq_len(length(args))) {
  eval(parse(text = args[[i]]))
}

# # Rprof("out.out")
# # library(numbat)
# library(future)
# plan(multisession)
# options(future.globals.maxSize= 3000*1024^2)
# library(beepr)
# library(progressr)
# handlers(global = TRUE)
# handlers("progress", "beepr")
# devtools::load_all()

library(numbat)
library(parallel)
library(Seurat)
library(readr)
library(magrittr)
# library(dtplyr)
conflicted::conflict_prefer("rowSums", "Matrix")

# debug(run_group_hmms)
# debug(get_exp_post)
# # debug(get_exp_sc)
# debug(find_common_diploid)
# debug(make_group_bulks)
# debug(run_group_hmms)
# # debug(get_exp_likelihoods)
# # debug(plot_psbulk)
# # debug(run_numbat)


matrix_dir = "/dataVolume/storage/single_cell_projects/resources/collin_et_al_2021_proj/output/cellranger/SRR13884248/outs/filtered_feature_bc_matrix/"

ncores = "4"
out_dir = "/dataVolume/storage/single_cell_projects/resources/collin_et_al_2021_proj/output/numbat/SRR13884248/"

allele_df_file = "/dataVolume/storage/single_cell_projects/resources/collin_et_al_2021_proj/output/numbat/SRR13884248_allele_counts.tsv.gz"

allele_df = data.table::fread(allele_df_file, sep = "\t") %>% 
  tidyr::drop_na(CHROM) %>%
  # dplyr::mutate(CHROM = as.factor(CHROM)) %>% 
  identity()

count_mat <- Seurat::Read10X(matrix_dir)

use_loh = NULL

# normal_reference_mat <- "~/single_cell_projects/resources/collin_et_al_2021_proj/output/numbat/count_mat.rds"
# 
# normal_reference_mat = readRDS(normal_reference_mat)
# 
# sridhar_seu <- readRDS("/dataVolume/storage/scEiad/sridhar_2020_seu.rds")
# 
# sridhar_cone_seu <- sridhar_seu[,sridhar_seu$CellType_predict == "Cone"]
# 
# ref_count_mat <- GetAssayData(sridhar_cone_seu, slot = "counts") %>% 
# 	as.data.frame() %>% 
# 	tibble::rownames_to_column("ensgene") %>% 
# 	dplyr::left_join(annotables::grch38[c("ensgene", "symbol")], by = "ensgene") %>% 
# 	dplyr::distinct(symbol, .keep_all = TRUE) %>% 
# 	dplyr::filter(!is.na(symbol)) %>% 
# 	tibble::column_to_rownames("symbol") %>% 
# 	# as.matrix() %>%
# 	identity()
# 
# normal_cell_annot <- 
# 	sridhar_cone_seu@meta.data[c("CellType_predict")] %>% 
# 	tibble::rownames_to_column("cell") %>% 
# 	dplyr::rename(group = CellType_predict)
# 	
# ref_internal = numbat::aggregate_counts(ref_count_mat, normal_cell_annot)
# 
# saveRDS(ref_internal, "data/sridhar_normal_ref.rds")

ref_internal <- readRDS("~/single_cell_projects/resources/collin_et_al_2021_proj/data/sridhar_normal_ref.rds")

# bulk = numbat:::get_bulk(
# 	count_mat = count_mat,
# 	lambdas_ref = ref_internal,
# 	df_allele = allele_df,
# 	gtf = gtf_hg38,
# 	genetic_map = genetic_map_hg38
# )
# 
# segs_loh = bulk %>% numbat:::detect_clonal_loh(t = 1e-4)

segs_loh = NULL
# run ------------------------------
# lambdas_ref = ref_hca
# lambdas_ref = readRDS("output/numbat/SRR13633761_ref.rds")

# debug(numbat::annot_consensus)
# debug(numbat::find_common_diploid)

out = run_numbat(
  count_mat = count_mat, # gene x cell integer UMI count matrix done
	lambdas_ref = ref_internal, # reference expression profile, a gene x cell type normalized expression level matrix done
	df_allele = allele_df, # allele dataframe generated by pileup_and_phase script done
	gtf = gtf_hg38, # provided upon loading the package done
	genetic_map = genetic_map_hg38, # provided upon loading the package done
	min_cells = 2,
	t = 1e-3,
	max_iter = 2,
	min_LLR = 10,
  max_entropy = 0.9,
  segs_loh = segs_loh,
	# min_LLR = 50,
	init_k = 3,
	ncores = as.integer(ncores),
	plot = TRUE,
	out_dir = out_dir, 
  tau = 0.3
)

